src/System.Text.JsonLab/System/Text/Json/Utf8JsonReaderStream.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.IO;
+
+namespace System.Text.JsonLab
+{
+    public ref struct Utf8JsonReaderStream
+    {
+        private Utf8JsonReader _jsonReader;
+        private Span<byte> _span;
+        private Stream _stream;
+        private byte[] _buffer;
+        private bool _isFinalBlock;
+        private long _consumed;
+
+        const int FirstSegmentSize = 1_024; //TODO: Is this necessary?
+        const int StreamSegmentSize = 4_096;
+
+        public JsonTokenType TokenType => _jsonReader.TokenType;
+        public ReadOnlySpan<byte> Value => _jsonReader.Value;
+        public long Consumed => _consumed + _jsonReader.Consumed;
+
+        public Utf8JsonReaderStream(Stream jsonStream)
+        {
+            if (!jsonStream.CanRead)
+                JsonThrowHelper.ThrowArgumentException("Stream must be readable");
+
+            _buffer = ArrayPool<byte>.Shared.Rent(FirstSegmentSize);
+            int numberOfBytes = jsonStream.Read(_buffer, 0, FirstSegmentSize);
+            _span = _buffer.AsSpan(0, numberOfBytes);
+            _stream = jsonStream;
+
+            _isFinalBlock = numberOfBytes < FirstSegmentSize;
```

Comment:
Read can return fewer bytes than were requested... the only way to know that there's no more data is if Read returns 0.

Reply:
Yes, I know. If Read returns less than FirstSegmentSize, _isFinalBlock is true.

If Read returns 0, we call the JsonReader with empty input, isFinalBlock = true.

Where does this break?

Reply:
> If Read returns less than FirstSegmentSize, _isFinalBlock is true. Where does this break?

Read could return 1 byte, and it wouldn't necessarily be the final block.  There could still be tons of data left, it just wasn't available yet when you issued the read.

Reply:
Ah I see. Hmm... that could be concern when setting isFinalBlock. I suppose I can set it to Read == 0. Is there any approach that avoids the extra call to Read()?

Reply:
"the" extra call to Read? It wouldn't necessarily be just one, right?  Presumably you'd only know for sure whether it was the final buffer or not if you repeatedly read until either you filled the buffer or got back 0.  And even then, you wouldn't know for sure unless you got back 0 before filling the buffer, as it's possible you fill the buffer, and it turns out the stream contained exactly that number of bytes, so that it really was the final buffer, you just didn't know it yet.  And given that you need to be able to support that case, even with sequences, presumably the reader is content with being given an empty final block.  In which case you could just do the single initial read and set isFinalBlock to true iff Read returned 0, as you say.

Reply:
> "the" extra call to Read? It wouldn't necessarily be just one, right?

I meant the extra call to the JonsReader.Read(), not stream.Read(). Once there is no more data, you have to call the JsonReader one last time with empty data. I understand that there could be many calls to stream.Read().
Presumably, there would just be **one** call to JsonReader.Read() at the end with empty data. All other calls to JsonReader should have some data, right? 

https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.read?view=netcore-2.1
> zero (0) if the end of the stream has been reached.

Given the docs, stream.Read should only ever return 0 once.

> And given that you need to be able to support that case, even with sequences, presumably the reader is content with being given an empty final block

Yes, it is.

Reply:
> All other calls to JsonReader should have some data, right?

Yes.

> Given the docs, stream.Read should only ever return 0 once.

Stream.Read will return 0 at EOF, never before.  But once at EOF, if you call Read again, it should return 0 again.